File: 06e4b7-simplex-algorithm/evaluation/evaluation.py
import os
import sys
import json
import uuid
import platform
import subprocess
import argparse
from datetime import datetime
from pathlib import Path

# --- Helper functions (Enhanced with Git and OS info) ---

def generate_run_id():
    return uuid.uuid4().hex[:8]

def get_git_info():
    """Get git commit and branch information."""
    git_info = {"git_commit": "unknown", "git_branch": "unknown"}
    try:
        result = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            capture_output=True, text=True, timeout=5
        )
        if result.returncode == 0:
            git_info["git_commit"] = result.stdout.strip()[:8]

        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True, text=True, timeout=5
        )
        if result.returncode == 0:
            git_info["git_branch"] = result.stdout.strip()
    except Exception:
        pass
    return git_info

def get_environment_info():
    git = get_git_info()
    return {
        "python_version": platform.python_version(),
        "platform": platform.platform(),
        "os": platform.system(),
        "git_commit": git["git_commit"],
        "git_branch": git["git_branch"],
    }

def generate_output_path():
    """Generate output path in format: evaluation/reports/YYYY-MM-DD/HH-MM-SS/report.json"""
    now = datetime.now()
    date_str = now.strftime("%Y-%m-%d")
    time_str = now.strftime("%H-%M-%S")

    project_root = Path(__file__).parent.parent
    output_dir = project_root / "evaluation" / "reports" / date_str / time_str
    output_dir.mkdir(parents=True, exist_ok=True)

    return output_dir / "report.json"

def parse_pytest_verbose_output(output):
    """Parse pytest verbose output to extract individual test results."""
    tests = []
    lines = output.split('\n')
    for line in lines:
        line = line.strip()
        if '::' in line:
            outcome = None
            if ' PASSED' in line: outcome = "passed"
            elif ' FAILED' in line: outcome = "failed"
            elif ' ERROR' in line: outcome = "error"
            elif ' SKIPPED' in line: outcome = "skipped"

            if outcome:
                nodeid = line.split(' ')[0]
                tests.append({
                    "nodeid": nodeid,
                    "name": nodeid.split("::")[-1],
                    "outcome": outcome,
                })
    return tests

def run_evaluation_tests(tests_dir):
    """Runs pytest and returns the structured data you requested."""
    cmd = [sys.executable, "-m", "pytest", str(tests_dir), "-v", "--tb=short"]

    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
        stdout = result.stdout
        stderr = result.stderr

        tests = parse_pytest_verbose_output(stdout)

        summary = {
            "total": len(tests),
            "passed": sum(1 for t in tests if t["outcome"] == "passed"),
            "failed": sum(1 for t in tests if t["outcome"] == "failed"),
            "errors": sum(1 for t in tests if t["outcome"] == "error"),
            "skipped": sum(1 for t in tests if t["outcome"] == "skipped"),
        }

        return {
            "success": result.returncode == 0,
            "exit_code": result.returncode,
            "tests": tests,
            "summary": summary,
            "stdout": stdout,
            "stderr": stderr
        }
    except Exception as e:
        return {
            "success": False,
            "exit_code": 1,
            "tests": [],
            "summary": {"total": 0, "passed": 0, "failed": 0, "errors": 1, "skipped": 0},
            "stdout": "",
            "stderr": str(e)
        }

# --- Main Evaluation Logic ---

def main():
    parser = argparse.ArgumentParser(description="Run evaluation")
    parser.add_argument("--output", type=str, default=None, help="Output JSON file path")
    args = parser.parse_args()

    # 1. Metadata
    run_id = generate_run_id()
    started_at = datetime.now()

    # 2. Paths
    project_root = Path(__file__).parent.parent
    tests_dir = project_root / "tests"

    # 3. Run Tests
    detailed_results = run_evaluation_tests(tests_dir)

    # 4. Finalize Timing
    finished_at = datetime.now()
    duration = (finished_at - started_at).total_seconds()

    # 5. Construct the specific JSON format
    report = {
        "run_id": run_id,
        "started_at": started_at.isoformat(),
        "finished_at": finished_at.isoformat(),
        "duration_seconds": round(duration, 4),
        "environment": get_environment_info(),
        "before": None,
        "after": detailed_results,
        "comparison": {
            "passed_gate": detailed_results["success"],
            "improvement_summary": "Evaluated implementation against provided test suite."
        },
        "success": detailed_results["success"],
        "error": None if detailed_results["success"] else "Tests failed in the 'after' implementation."
    }

    # 6. Save Report
    if args.output:
        output_path = Path(args.output)
    else:
        output_path = generate_output_path()

    output_path.parent.mkdir(parents=True, exist_ok=True)

    with open(output_path, "w") as f:
        json.dump(report, f, indent=2)

    print(f"\n✅ Report saved to: {output_path}")
    print(f"Summary: {detailed_results['summary']['passed']}/{detailed_results['summary']['total']} passed")

    sys.exit(0 if detailed_results["success"] else 1)

if __name__ == "__main__":
    main()

File: f9b421-Order_Service_Database_Fix/evaluation/evaluation.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const { spawn } = require("child_process");
const os = require("os");
const crypto = require("crypto");

// --- CONFIGURATION ---
const ROOT_DIR = path.resolve(__dirname, "..");

// FIX: Configuration now maps specific test files to specific scenarios
const TARGETS = {
  before: {
    // Matches: tests/test-before-repository-suite.js
    testFile: path.join(ROOT_DIR, "tests", "test-before-repository-suite.js"),
    label: "Before (Unoptimized)",
  },
  after: {
    // Matches: tests/test-after-repository-suite.js
    testFile: path.join(ROOT_DIR, "tests", "test-after-repository-suite.js"),
    label: "After (Optimized)",
  },
};

// --- HELPER FUNCTIONS ---

function generateRunId() {
  return crypto.randomBytes(4).toString("hex");
}

function getGitInfo() {
  const info = { commit: "unknown", branch: "unknown" };
  try {
    info.commit = require("child_process")
      .execSync("git rev-parse HEAD", { timeout: 1000, stdio: "pipe" })
      .toString()
      .trim()
      .substring(0, 8);
    info.branch = require("child_process")
      .execSync("git rev-parse --abbrev-ref HEAD", {
        timeout: 1000,
        stdio: "pipe",
      })
      .toString()
      .trim();
  } catch (e) {
    // Git not available, ignore silently
  }
  return info;
}

function getEnvironmentInfo() {
  const git = getGitInfo();
  return {
    node_version: process.version,
    platform: os.platform(),
    release: os.release(),
    arch: os.arch(),
    hostname: os.hostname(),
    git_commit: git.commit,
    git_branch: git.branch,
  };
}

/**
 * Runs the specific test suite file
 */
function runTestSuite(targetLabel, testFilePath) {
  return new Promise((resolve) => {
    // FIX: Added logging so you can see progress immediately
    console.log(`\n▶ Running: ${targetLabel}`);
    console.log(`  File: ${path.relative(ROOT_DIR, testFilePath)}`);

    const startTime = Date.now();

    // Spawn the specific test file
    const child = spawn("node", [testFilePath], {
      cwd: ROOT_DIR,
      env: process.env, // Pass current env
      stdio: "pipe",
    });

    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (data) => {
      const str = data.toString();
      stdout += str;
      // Optional: Uncomment the next line if you want to see test logs in real-time
      // process.stdout.write(str);
    });

    child.stderr.on("data", (data) => {
      const str = data.toString();
      stderr += str;
      // FIX: Always print errors so you know if files are missing
      process.stderr.write(str);
    });

    child.on("close", (code) => {
      const duration = (Date.now() - startTime) / 1000;

      // If the process failed, we should know why immediately
      if (code !== 0) {
        console.log(`  ⚠️  Process exited with code ${code}`);
        if (!stderr && !stdout) console.log("  ⚠️  No output received.");
      }

      const results = parseTestOutput(stdout);

      resolve({
        success: code === 0,
        exit_code: code,
        duration_seconds: duration,
        results: results,
        raw_output: stdout,
        raw_error: stderr,
      });
    });
  });
}

function parseTestOutput(output) {
  const tests = [];
  const lines = output.split("\n");
  // Adjust these regexes depending on how your specific test files output data
  const testLineRegex = /^(.*?)\s\.\.\.\s(✅ PASS|❌ FAIL|passed|failed)/i;
  const summaryRegex = /TEST SUMMARY: (\d+)\/(\d+) Passed/i;

  let totalParsed = 0;
  let passedParsed = 0;

  for (const line of lines) {
    const match = line.match(testLineRegex);
    if (match) {
      tests.push({
        name: match[1].trim(),
        outcome: match[2].toLowerCase().includes("pass") ? "passed" : "failed",
      });
    }

    const sumMatch = line.match(summaryRegex);
    if (sumMatch) {
      passedParsed = parseInt(sumMatch[1], 10);
      totalParsed = parseInt(sumMatch[2], 10);
    }
  }

  // Fallback if regex didn't catch specific summary lines (common in custom test runners)
  if (totalParsed === 0 && tests.length > 0) {
    totalParsed = tests.length;
    passedParsed = tests.filter((t) => t.outcome === "passed").length;
  }

  return {
    tests,
    summary: {
      total: totalParsed,
      passed: passedParsed,
      failed: totalParsed - passedParsed,
    },
  };
}

function generateReportPath() {
  const now = new Date();
  const dateStr = now.toISOString().split("T")[0];
  const timeStr = now.toTimeString().split(" ")[0].replace(/:/g, "-");

  const reportDir = path.join(ROOT_DIR, "evaluation");

  // Ensure directory exists
  if (!fs.existsSync(reportDir)) {
    fs.mkdirSync(reportDir, { recursive: true });
  }

  return path.join(reportDir, "report.json");
}

// --- MAIN EXECUTION ---

async function main() {
  const runId = generateRunId();
  const startedAt = new Date();

  try {
    // Run Before Suite
    const beforeResult = await runTestSuite(
      TARGETS.before.label,
      TARGETS.before.testFile
    );

    // Run After Suite
    const afterResult = await runTestSuite(
      TARGETS.after.label,
      TARGETS.after.testFile
    );

    const finishedAt = new Date();

    const comparison = {
      before_success: beforeResult.success,
      after_success: afterResult.success,
      before_score: `${beforeResult.results.summary.passed}/${beforeResult.results.summary.total}`,
      after_score: `${afterResult.results.summary.passed}/${afterResult.results.summary.total}`,
      performance_impact: {
        before_duration: beforeResult.duration_seconds,
        after_duration: afterResult.duration_seconds,
        diff: (
          afterResult.duration_seconds - beforeResult.duration_seconds
        ).toFixed(3),
      },
    };

    const report = {
      run_id: runId,
      started_at: startedAt.toISOString(),
      finished_at: finishedAt.toISOString(),
      environment: getEnvironmentInfo(),
      comparison: comparison,
      details: {
        before: beforeResult,
        after: afterResult,
      },
    };

    const reportPath = generateReportPath();
    fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));

    console.log("\n✅ Evaluation Completed successfully!");
    console.log(`📄 Report saved to: ${reportPath}`);

    // Check if we actually ran tests
    if (beforeResult.results.summary.total === 0) {
      console.warn(
        "\n⚠️  WARNING: No tests were detected in the 'Before' suite. Check file paths."
      );
    }

    process.exit(afterResult.success && beforeResult.success ? 0 : 1);
  } catch (error) {
    console.error("\n❌ CRITICAL ERROR:", error);
    process.exit(1);
  }
}

main();
